SQL: Fix FORMAT function to better comply with Microsoft SQL Server specification

luigidellaquila opened the pull request:
Based on the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-format), `FORMAT()` function should comply with [Microsoft SQL Server Format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) specification. 

Current implementation uses Java DateTimeFormat as it is, only with replacement of basic pattern characters that are slightly different between the two specs, but it's not fully compliant.

The goal of this PR is to better translate Microsoft datetime patterns to the equivalent in Java, and addresses the cases that are not currently covered, ie.
- allowed characters in Microsoft spec that are not allowed in Java are now quoted and returned correctly
- quoting syntax (that is different in the two specs) is properly translated
- some further fixes to patterns translation (eg. `Y` and `y` in Java are equivalent, while in MS spec `Y` is just a normal character, not interpreted as a pattern) 


Fixes #66560

luigidellaquila committed:
Fix SQL FORMAT function to comply with Microsoft SQL Server spec

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

luigidellaquila committed:
Update docs/changelog/86225.yaml

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

elasticmachine committed:
Merge branch 'master' into fix/format_date_fix

pugnascotia commented on x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec:
```diff
@@ -136,39 +136,39 @@ SELECT ISO_DAY_OF_WEEK(birth_date) AS d, SUM(salary) s FROM test_emp GROUP BY d
 
       d:i      |      s:l
 ---------------+---------------
-7              |386466       
```
@luigidellaquila this is probably due to `.editorconfig`. You could add some further config to it, which worked when I tested it:

```diff
diff --git a/.editorconfig b/.editorconfig
index 92111b6c074..00dc2f6fa83 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -225,3 +225,6 @@ indent_size = 2
 
 [*.{xsd,xml}]
 indent_size = 4
+
+[*.{csv,sql}-spec]
+trim_trailing_whitespace = false
```

luigidellaquila replied:
Thank you very much @pugnascotia, I'll try it

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessorTests.java:
```diff
@@ -188,45 +177,44 @@ public void testFormatting() {
         );
         assertEquals(
             "2019-09-04 04:10:37.12345678 AM",
-            new Format(Source.EMPTY, dateTime, l("YYYY-MM-dd HH:mm:ss.ffffffff t"), zoneId).makePipe().asProcessor().process(null)
```
I would also add a test with `YYYY` to show that the MS pattern is case-sensitive.

astefan reviewed (commented):
Left some comments.
Also, I suggest changing the title of the PR to `Fix FORMAT function to better comply with Microsoft SQL Server specification`. Not all the characteristics of MS format are supported.

luigidellaquila commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:
```diff
@@ -67,6 +86,105 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
+            StringBuilder partialQuotedString = new StringBuilder();
+
+            boolean originQuoting = false;
+            boolean targetQuoting = false;
```
I see, the naming is bad and misleading.
What I mean here is:
- `originQuoting`: the character in the original (MS) pattern was intended to be quoted (eg. previous char was a backslash or is inside a quoting block).
- `targetQuoting`: last character in the generated (java) pattern has to be quoted, so I'm building a portion of pattern that will then be quoted. Unfortunately I cannot quote characters one by one, I have to accumulate portions of string that have to be quoted and then quote them properly

I'll try to find better names

Commit:
Fix based on review comments
- code refactoring
- further unit tests
- fix .editorconfig to avoid trim of trailing whitespaces in csv-spec files

Luegg reviewed (commented):
Are there more breaking changes apart from `Y` vs. `y`?

For `Y` vs `y` I would suggest to keep supporting upper case `Y` and potentially emit a deprecation warning instead of breaking current behavior. Since even the docs uses `Y` in the example it seems quite likely that users also do so.

luigidellaquila commented:
@elasticmachine run CLA

luigidellaquila commented:
@Luegg yes, there are many breaking changes here, like A, N, G, VV... 
Actually, with the original implementation, only nine MS patterns were translated to Java (see [diff](https://github.com/elastic/elasticsearch/pull/86225/files#diff-2f663032b1c27e10472d85bea179691ebadb5a537fc7dfd7f0684e51fd03c262R66)), all the rest of the characters were just copied and interpreted with Java semantics. As you can [see](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html), there are maybe ten or more characters that Java DateTimeFormatters interprets as valid date patterns and that MS just ignores.

IMHO it could still make sense to treat Y as an exception, since it's pretty common, probably we should not maintain the compatibility with all the other cases.

astefan commented:
> For `Y` vs `y` I would suggest to keep supporting upper case `Y` and potentially emit a deprecation warning instead of breaking current behavior. Since even the docs uses `Y` in the example it seems quite likely that users also do so.

I have mixed feelings about this. This is a bug and should be treated like it in my opinion: changing the output means fixing a faulty behavior in the first place. On the other hand, we could do as you suggest and be lenient towards such changes, but this could introduce a precedent: every bug fix should also consider any (if it wouldn't be a bug fix) introduced breaking changes and add deprecation warnings and so on. @bpintea @costin wdyt?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:
```diff
@@ -67,6 +100,99 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
+                        originalCharacterQuoted = false;
+                        lastTargetCharacterQuoted = true;
+                        partialQuotedString.append(c);
+                        continue;
```
Please use `else if` instead of `continue` through-out this method.

costin reviewed (approved):
LGTM.
Left some comments around the coding style to replace the majority of `if .. continue` with `else if`.
Regarding the changes, I don't think we need to be lenient - this function is specific to SQL Server and should act accordingly. The fact that `Y` has been supported instead of `y` is a bug - if folks want to use the Java formatter rules, they can do so using the regular `DATE_FORMATTER` or simply change their queries to use `y` instead. It is annoying however it has been a bug that affected also our documentation.
In short, I would just apply the changes without any special treatment of the previous behavior.

Luegg commented:
Interesting, I wasn't aware that we support 3 different formats for date formatting and assumed that's the main one.

In any case, there is one more occurrence of the upper case `Y` in the docs:

<img width="751" alt="image" src="https://user-images.githubusercontent.com/1016746/165794612-9de79a5e-b576-4f48-8420-cfd42b50003c.png">

luigidellaquila committed:
Refactor DateTimeFormatProcessor and fix docs

astefan reviewed (approved):
LGTM

bpintea commented:
I'm wondering just how close to the MS specification we will want to get?
I'm thinking of the treatment of `%`, which has this [single custom format specifier](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings#UsingSingleSpecifiers) semantic, or single `"` characters, which are just piped through by MS's `FORMAT`, but swallowed by current implementation, or double tick `''`, that will be piped through by MS's `FORMAT`, but treated as a single `'` by DateTimeFormatter.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:
```diff
@@ -67,6 +100,95 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
+                                // and then translate the pattern
+                                result.append(item[1]);
+                                characterProcessed = true;
+                                i += (fragmentLength - 1); // fast-forward, because the replaced pattern could be longer than one character
```
Nit: the parentheses shouldn't be needed. Just in case any other changes will be needed.
Also, I'm generally in favour of using a while loop and explicitly incrementing by needed amount (vs. decrementing due to for).

bpintea reviewed (approved):
LGTM

luigidellaquila commented:
> Are there more breaking changes apart from Y vs. y?

I saw a recent email from the Breaking Changes Committee, I was wondering if this could be relevant and if it makes sense to submit the case. WDYT?
